File: src/main/java/main/commands/slashcommands/commands/management/permissions/RemoveDJCommand.java
package main.commands.slashcommands.commands.management.permissions;

import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.constants.Permission;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.component.interactions.AbstractSlashCommand;
import main.utils.json.permissions.PermissionsConfig;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.script.ScriptException;
import java.util.List;

public class RemoveDJCommand extends AbstractSlashCommand implements ICommand {
    private final Logger logger = LoggerFactory.getLogger(RemoveDJCommand.class);

    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final List<String> args = ctx.getArgs();
        final Message msg = ctx.getMessage();
        final Guild guild = ctx.getGuild();

        EmbedBuilder eb;

        if (!GeneralUtils.hasPerms(guild, ctx.getMember(), Permission.ROBERTIFY_ADMIN)) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.INSUFFICIENT_PERMS_NO_ARGS);
            msg.replyEmbeds(eb.build()).queue();
            return;
        }

        if (args.isEmpty()) {
            eb = RobertifyEmbedUtils.embedMessage(guild, "You must provide a role to remove as a DJ role!");
            msg.replyEmbeds(eb.build()).queue();
            return;
        }

        String id = GeneralUtils.getDigitsOnly(args.get(0));

        if (!GeneralUtils.stringIsID(id)) {
            msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "That is an invalid ID!\nMake sure to either **mention** a user or provide their **ID**")
                            .setImage("https://i.imgur.com/wa8CjnJ.png")
                            .build())
                    .queue();
            return;
        }

        Role role = guild.getRoleById(id);
        User user = GeneralUtils.retrieveUser(id);

        if (role == null && user == null) {
            eb = RobertifyEmbedUtils.embedMessage(guild, "Please provide a valid role/user ID!");
            msg.replyEmbeds(eb.build()).queue();
            return;
        }

        if (user == null)
            msg.replyEmbeds(handleDJRemove(guild, role).build()).queue();
        else
            msg.replyEmbeds(handleDJRemove(guild, user).build()).queue();
    }

    private EmbedBuilder handleDJRemove(Guild guild, Role role) {
        PermissionsConfig permissionsConfig = new PermissionsConfig(guild);
        try {
            permissionsConfig.removeRoleFromPermission(role.getIdLong(), Permission.ROBERTIFY_DJ);
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.PermissionsMessages.DJ_REMOVED, Pair.of("{mentionable}", role.getAsMention()));
        } catch (IllegalAccessException e) {
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.PermissionsMessages.NOT_DJ, Pair.of("{djType}", "role"));
        } catch (Exception e) {
            logger.error("[FATAL ERROR] An unexpected error occurred!", e);
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.UNEXPECTED_ERROR);
        }
    }

    private EmbedBuilder handleDJRemove(Guild guild, User user) {
        PermissionsConfig permissionsConfig = new PermissionsConfig(guild);
        try {
            permissionsConfig.removePermissionFromUser(user.getIdLong(), Permission.ROBERTIFY_DJ);
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.PermissionsMessages.DJ_REMOVED, Pair.of("{mentionable}", user.getAsMention()));
        } catch (IllegalArgumentException e) {
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.PermissionsMessages.NOT_DJ, Pair.of("{djType}", "user"));
        } catch (Exception e) {
            logger.error("[FATAL ERROR] An unexpected error occurred!", e);
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.UNEXPECTED_ERROR);
        }
    }

    @Override
    public String getName() {
        return "removedj";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+ GeneralUtils.listToString(getAliases())+"`\n" +
                "Remove DJ privileges from a specific role\n\n" +
                "Usage: `"+ prefix +"remove <@role|@user>`";
    }

    @Override
    public List<String> getAliases() {
        return List.of("rdj");
    }

    @Override
    protected void buildCommand() {
        setCommand(
                getBuilder()
                        .setName("removedj")
                        .setDescription("Remove a role/user as a DJ")
                        .addSubCommands(
                                SubCommand.of(
                                        "role",
                                        "Remove a role as a DJ",
                                        List.of(CommandOption.of(
                                                OptionType.ROLE,
                                                "role",
                                                "The role to remove as a DJ",
                                                true
                                        ))
                                ),
                                SubCommand.of(
                                        "user",
                                        "Remove a user as a DJ",
                                        List.of(CommandOption.of(
                                                OptionType.USER,
                                                "user",
                                                "The user to remove as a DJ",
                                                true
                                        ))
                                )
                        )
                        .setAdminOnly()
                        .build()
        );
    }

    @Override
    public String getHelp() {
        return null;
    }

    @Override
    public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event) {
        if (!checks(event)) return;

        switch  (event.getSubcommandName()) {
            case "role" -> {
                var role = event.getOption("role").getAsRole();
                event.replyEmbeds(handleDJRemove(event.getGuild(), role).build())
                        .setEphemeral(RobertifyEmbedUtils.getEphemeralState(event.getChannel().asGuildMessageChannel()))
                        .queue();
            }
            case "user" -> {
                var user = event.getOption("user").getAsUser();
                event.replyEmbeds(handleDJRemove(event.getGuild(), user).build())
                        .setEphemeral(RobertifyEmbedUtils.getEphemeralState(event.getChannel().asGuildMessageChannel()))
                        .queue();
            }
        }
    }
}

File: src/main/kotlin/main/commands/slashcommands/management/permissions/RemoveDJCommand.kt
package main.commands.slashcommands.management.permissions

import main.commands.slashcommands.SlashCommandManager.getRequiredOption
import main.constants.RobertifyPermission
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.json.permissions.PermissionsConfig
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.PermissionsMessages
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.IMentionable
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType

class RemoveDJCommand : AbstractSlashCommand(SlashCommand(
    name = "removedj",
    description = "Remove a user or role as a DJ in this server.",
    options = listOf(
        CommandOption(
            type = OptionType.MENTIONABLE,
            name = "target",
            description = "The user or role to remove as a DJ."
        )
    ),
    adminOnly = true
)) {

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        val mentionable = event.getRequiredOption("target").asMentionable
        event.replyEmbed { handleRemoveDJ(event.guild!!, mentionable) }
            .queue()
    }

    private suspend fun handleRemoveDJ(guild: Guild, mentionable: IMentionable): MessageEmbed {
        val config = PermissionsConfig(guild)

        return when {
            mentionable.asMention.startsWith("<@&") -> {
                try {
                    config.removeRoleFromPermission(mentionable.idLong, RobertifyPermission.ROBERTIFY_DJ)
                    RobertifyEmbedUtils.embedMessage(
                        guild,
                        PermissionsMessages.DJ_REMOVED,
                        Pair("{mentionable}", mentionable.asMention)
                    ).build()
                } catch (e: IllegalAccessException) {
                    RobertifyEmbedUtils.embedMessage(
                        guild,
                        PermissionsMessages.MENTIONABLE_NEVER_HAD_PERMISSION,
                        Pair("{mentionable}", mentionable.asMention),
                        Pair("{permission}", RobertifyPermission.ROBERTIFY_DJ.name)
                    ).build()
                }
            }

            mentionable.asMention.startsWith("<@") -> {
                if (!config.userHasPermission(mentionable.idLong, RobertifyPermission.ROBERTIFY_DJ))
                    return RobertifyEmbedUtils.embedMessage(
                        guild,
                        PermissionsMessages.MENTIONABLE_NEVER_HAD_PERMISSION,
                        Pair("{mentionable}", mentionable.asMention),
                        Pair("{permission}", RobertifyPermission.ROBERTIFY_DJ.name)
                    ).build()

                config.removePermissionFromUser(mentionable.idLong, RobertifyPermission.ROBERTIFY_DJ)
                RobertifyEmbedUtils.embedMessage(
                    guild,
                    PermissionsMessages.DJ_REMOVED,
                    Pair("{mentionable}", mentionable.asMention)
                ).build()
            }

            else -> RobertifyEmbedUtils
                .embedMessage(guild, GeneralMessages.INVALID_ARGS)
                .build()
        }
    }
}